modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -145,13 +156,266 @@ class SoftMaxSubgraph : public Subgraph
+    {
+        return Subgraph::match(net, nodeId, matchedNodesIds, targetNodesIds);
+    }
+
+    void finalize(const Ptr<ImportGraphWrapper>& net,
+                          const Ptr<ImportNodeWrapper>& fusedNode,
+                          std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        opencv_onnx::AttributeProto* attr = node->add_attribute();
+        attr->set_name("scales");
+        attr->add_ints(1);
+        attr->add_ints(1);
+        attr->add_ints(scaleH);
+        attr->add_ints(scaleW);
+        //remove empty constant node
+        if(emptyNodeId != -1)
+            net->removeNode(emptyNodeId);
+    }
+
+protected:
+    int scaleH, scaleW, emptyNodeId;
+};
+
+class ExtractScalesSubgraph : public FuseUpsampleSubgraph
+{
+public:
+    ExtractScalesSubgraph()
+    {
+        int input = addNodeToMatch("");
+
+        int indexH = addNodeToMatch("Constant");
+        int shape1 = addNodeToMatch("Shape", input);
+        int gather1 = addNodeToMatch("Gather", shape1, indexH);
+        int castG1 = addNodeToMatch("Cast", gather1);
+        int scaleH = addNodeToMatch("Constant");
+        int mul1 = addNodeToMatch("Mul", castG1, scaleH);
+        int castM1 = addNodeToMatch("Cast", mul1);
+        int floor1 = addNodeToMatch("Floor", castM1);
+
+        int indexW = addNodeToMatch("Constant");
+        int shape2 = addNodeToMatch("Shape", input);
+        int gather2 = addNodeToMatch("Gather", shape2, indexW);
+        int castG2 = addNodeToMatch("Cast", gather2);
+        int scaleW = addNodeToMatch("Constant");
+        int mul2 = addNodeToMatch("Mul", castG2, scaleW);
+        int castM2 = addNodeToMatch("Cast", mul2);
+        int floor2 = addNodeToMatch("Floor", castM2);
+
+        int unsqueeze1 = addNodeToMatch("Unsqueeze", floor1);
+        int unsqueeze2 = addNodeToMatch("Unsqueeze", floor2);
+        addNodeToMatch("Concat", unsqueeze1, unsqueeze2);
+
+        setFusedNode("Scales", input);
+    }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                    std::vector<int>& matchedNodesIds,
+                    std::vector<int>& targetNodesIds) CV_OVERRIDE
+    {
+        if (FuseUpsampleSubgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
+        {
+            std::vector<cv::Mat> nodes_attributes;
+            for (int i = matchedNodesIds.front(); i < matchedNodesIds.back(); i++)
+            {
+                if (std::find(matchedNodesIds.begin(), matchedNodesIds.end(), i) == matchedNodesIds.end())
+                {
+                    opencv_onnx::NodeProto* constant_node = net->getNode(i).dynamicCast<ONNXNodeWrapper>()->node;
+                    opencv_onnx::TensorProto tensor_proto = constant_node->attribute(0).t();
+                    nodes_attributes.push_back(getMatFromTensor(tensor_proto));
+                }
+            }
```

Comment:
What this loop actually does?

Reply:
Sequence of our subgraph is

`Shape Constant Gather Cast Constant Mul Cast Floor Shape Constant Gather Cast Constant Mul Cast Floor Unsqueeze Unsqueeze`

While matchedNodesIds contain vector ids of nodes 

`Shape Gather Cast Mul Cast Floor Shape Gather Cast Mul Cast Floor Unsqueeze Unsqueeze`
i.e. without constant. 
so this loop is searching  for consant nodes that are missing from matchedNodeIds because we need to extract data from those constant nodes.